google/cloud/internal/rest_opentelemetry.cc

```diff
@@ -0,0 +1,68 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#if GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+#include "google/cloud/internal/rest_opentelemetry.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+#include "google/cloud/internal/opentelemetry.h"
+#include "google/cloud/internal/opentelemetry_options.h"
+#include "google/cloud/options.h"
+#include "absl/strings/match.h"
+#include <opentelemetry/context/propagation/global_propagator.h>
+#include <opentelemetry/context/propagation/text_map_propagator.h>
+#include <opentelemetry/trace/provider.h>
+#include <opentelemetry/trace/semantic_conventions.h>
+#include <opentelemetry/trace/span_metadata.h>
+#include <opentelemetry/trace/span_startoptions.h>
+
+namespace google {
+namespace cloud {
+namespace rest_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> MakeSpanHttp(
+    RestRequest const& request, opentelemetry::nostd::string_view method) {
+  namespace sc = opentelemetry::trace::SemanticConventions;
+  opentelemetry::trace::StartSpanOptions options;
+  options.kind = opentelemetry::trace::SpanKind::kClient;
+  auto span =
+      google::cloud::internal::GetTracer(
+          google::cloud::internal::CurrentOptions())
+          ->StartSpan(absl::StrCat("HTTP/", absl::string_view{method.data(),
+                                                              method.size()}),
+                      {{sc::kNetTransport, sc::NetTransportValues::kIpTcp},
+                       {sc::kHttpMethod, method},
+                       {sc::kHttpUrl, request.path()}},
+                      options);
+  for (auto const& kv : request.headers()) {
```

Comment:
comment: It is nice that we know all of the headers are present by the time the `RestRequest` gets to the `RestClient`. It makes me wonder if I could've / should've tried to do tracing in an extension of the gRPC `StubInterface` instead of our stubs. Also, my guess is that a REST tracing stub (for fully generated REST services) will not be necessary.

Reply:
> Also, my guess is that a REST tracing stub (for fully generated REST services) will not be necessary.

I am not so sure.  Once still may want to know the intent of a particular HTTP request, e.g., a `GET` request may return the metadata for a GCS object, or the full data, depending on the `alt=media` parameter.

There is also the need to capture the retry span... and resume spans for downloads and uploads.

Reply:
I may take that shortcut to get tracing data in the current troubleshooting that I am involved with.